### ISSUE TITLE
RethinkDB is overall really really slow with a table containing millions of rows

### ISSUE THREAD

**Issue by thelinuxlich** [rethinkdb/rethinkdb#5204]
I had a cluster of three google cloud instances with 4-core, 16gb RAM, 200GB SSD disk and got this issue #5195 

Then I migrated the three machines to 16-core instances and created a test table with the same data as the one on issue #5195, but duplicated to millions(13k turned into 17kk)

Then I found this issue #5201, which is strange because you advertise that writes don't block reads and vice-versa.

Now I'm testing the big table:

```javascript
r.table("test_pageviews").count()
// took 5.5 seconds to return 17749703

r.table("test_pageviews").between([7,r.time(2015,12,1,"-02:00")],[7,r.now()],{index: "client_id_created_at"}).count()
// Never completed and also made two instances lock, so I had to restart them

```
All tables have 3 shards, 9 replicas. After restarting the locked instances, Some tables have 6 and others have 7 replicas and it doesn't return back to normal. I tried reconfiguring every table, but the issue stands. All instances are with idle CPU
I noticed one of the instances had 15gb RAM consumed of 16gb(cache is set to 10gb) and the others are consuming just 3gb, so I restarted the one almost going to swap.
Once it restarted, MAGIC! All tables are ready!

Tried the count with secondary index again, reads spike to 19k/second on the Data Explorer, cache goes to 33%(and stays that a long time after the query executed)
I doubt readMode: "outdated" will work a miracle here so I didn't test that.
With this result, I know that even if I made the joined session table a embedded document, it wouldn't help.

Now I tried to execute the original query of issue #5195. Reads spiked to 33k/second on the Data Explorer, all 16 cores of the instance I'm running the script spike to ~35%. The instance where I was running the script went down with this error "rethinkdb: Memory allocation failed. This usually means that we have run out of RAM. Aborting.", 10gb RAM weren't enough for those 2 million records.

**Comment by danielmewes** [comment]
We're investigating. There might be a bug...

**Comment by thelinuxlich** [comment]
Oh, one thing I missed to tell is that the ~18 million test table is consuming >70GB disk space

**Comment by thelinuxlich** [comment]
So I rebuilt my cluster now using local SSDs(170k IOPS) and rerun the same queries:

```
r.table("test_pageviews").count()
// oddly took 50 seconds to return 17749703

r.table("test_pageviews").between([7,r.time(2015,12,1,"-02:00")],[7,r.now()],{index: "client_id_created_at"}).count()
// 14 seconds
```

I also run the query from issue #5195, much smaller dataset, but using joins and group and it took roughly the same time as the old hardware.

**Comment by thelinuxlich** [comment]
Average iostat from the second query:

```
Device:            tps    MB_read/s    MB_wrtn/s    MB_read    MB_wrtn
nvme0n1        7257.00       344.57         0.00        344          0
sda               0.00         0.00         0.00          0          0
```

**Comment by thelinuxlich** [comment]
By the way, using a benchmark explained in google cloud docs, I can see the local ssd is reaching 640k IOPS

**Comment by thelinuxlich** [comment]
Changed io-threads to 128, nothing different. Also, now that I'm running only the second query, I can see that its average time is 50s

**Comment by danielmewes** [comment]
@thelinuxlich I suspect that it's now hitting the throughput limit. It appears to be doing ~345 MB/s. According to https://cloud.google.com/compute/docs/disks/#comparison_of_disk_types , Google gives you 1.77 MB/s per GB on NVMe storage. So for a 200 GB volume that would be 340 MB/s.

**Comment by thelinuxlich** [comment]
I get ~2500MB/s benchmarking the disk

**Comment by thelinuxlich** [comment]
By the way, Google local SSD has a fixed size of 375GB

**Comment by thelinuxlich** [comment]
I don't know if you are still considering this a bug, I'm very disappointed with the results. 6k IOPS to 640k IOPS and the query reduced from 150s to 50s.

**Comment by danielmewes** [comment]
Yeah it looks like this is just an i/o bottleneck, possibly an inefficiency with how well we utilize high-throughput i/o systems.
I'm going to remove the `bug` tag. We should still see if we can get more out of those disks by optimizing our access patterns.

**Comment by ChrisCates** [comment]
@thelinuxlich Me and @danielmewes were discussing this earlier.
I believe the solution is to convert the `.count()` operation to a HyperLogLog computation.

Once I clear up my client work, I'll start figuring it out how to implement natively with RethinkDB (I don't normally write C++ :scream:)

I'll check in later then *potentially* do a PR.

**Comment by deontologician** [comment]
Hyperloglog doesn't really seem like it would help with `count()`. It's meant to save space when counting unique items in a set, but when counting all items in a set it's even more space efficient to just use a counter.

**Comment by thelinuxlich** [comment]
@ChrisCates I think it's something related to RethinkDB not really using all the potential of speedier disks like non-volatile memory(nvme). On benchmarks I can reach 640k IOPS easily, but RethinkDB under heavy load peaks at 50k and average 30k IOPS.

**Comment by ChrisCates** [comment]
@deontologician @thelinuxlich Fair enough... :cry:, not sure how to resolve this.

**Comment by thomasmodeneis** [comment]
To get queries like this working I had to mirror rethinkdb vs elasticsearch and got me some speed 👍  

I could do this permanently and forget about this problem for a while, however not having reliable changefeeds is also a killer :|

To work around its only by usign two secondary indexes then the performance is better, actually is the only way to get any query some speed after some millions.

Cheers.

**Comment by mackeyja92** [comment]
I am seeing similar issues. I have a cluster of 6 nodes all with high performance SSDs, 128Gb ram, 40 logical cores of decent speed. I currently have a table with 300M ish documents in it when doing a filtered count on that table usually locks up the nodes for about 30 minutes bringing everything to a stall where all other queries stop processing on that table.

**Comment by osullivj** [comment]
I had a similar issue with slow .filter( ) results on a 5M doc table. I
rejigged my schema and query code to search on an indexed field and now
have fast queries. Using filter means pulling every record in the table;
indexed searches are much quicker. I guess it all comes down to avoiding
queries analogous to SQL's "like '%<searchTerm>%'.
